File: saucy-menu-backend/utils/audit.ts
import { db } from "../src/db";
import { AuditAction, AuditEntity, auditLogs } from "../src/db/schema";


interface LogAuditParams {
    entity: AuditEntity;
    entityId: string;
    action: AuditAction;
    performedBy?: string;
    before?: Record<string, unknown> | null;
    after?: Record<string, unknown> | null;
    metadata?: Record<string, unknown>;
    restaurantId?: string;
}

export async function logAudit(params: LogAuditParams) {
    const { entity, entityId, action, performedBy, before, after, metadata, restaurantId } = params;

    await db.insert(auditLogs).values({
        entity,
        entityId,
        action,
        performedBy: performedBy ?? null,
        before: before ?? null,
        after: after ?? null,
        metadata: metadata ?? null,
        restaurantId: restaurantId ?? null,
    });
}

// --- Convenience wrappers ---

export const audit = {
    created: ({ entity, entityId, after, performedBy, metadata, restaurantId }: { entity: AuditEntity, entityId: string, after: Record<string, unknown>, performedBy?: string, metadata?: Record<string, unknown>, restaurantId?: string }) =>
        logAudit({ entity, entityId, action: "CREATE", after, performedBy, metadata, restaurantId }),

    updated: ({ entity, entityId, before, after, performedBy, metadata, restaurantId }: { entity: AuditEntity, entityId: string, before: Record<string, unknown>, after: Record<string, unknown>, performedBy?: string, metadata?: Record<string, unknown>, restaurantId?: string }) =>
        logAudit({ entity, entityId, action: "UPDATE", before, after, performedBy, metadata, restaurantId }),

    deleted: ({ entity, entityId, before, performedBy, metadata, restaurantId }: { entity: AuditEntity, entityId: string, before: Record<string, unknown>, performedBy?: string, metadata?: Record<string, unknown>, restaurantId?: string }) =>
        logAudit({ entity, entityId, action: "DELETE", before, performedBy, metadata, restaurantId }),
};

File: saucy-menu-backend-go/internal/audit/audit.go
package audit

import (
	"context"
	"encoding/json"

	"github.com/jackc/pgx/v5/pgtype"
	"github.com/rs/zerolog/log"

	"github.com/restorefine-studios/saucy-menu-backend-go/internal/db/sqlc"
)

type Logger struct {
	q *sqlc.Queries
}

func New(q *sqlc.Queries) *Logger {
	return &Logger{q: q}
}

func (l *Logger) log(ctx context.Context, entity sqlc.AuditEntity, entityID, performedBy, restaurantID pgtype.UUID, action sqlc.AuditAction, before, after any) {
	beforeJSON := marshalJSON(before)
	afterJSON := marshalJSON(after)

	if err := l.q.InsertAuditLog(ctx, sqlc.InsertAuditLogParams{
		Entity:       entity,
		EntityID:     entityID,
		Action:       action,
		PerformedBy:  performedBy,
		Column5:      beforeJSON,
		Column6:      afterJSON,
		RestaurantID: restaurantID,
	}); err != nil {
		log.Error().Err(err).Str("entity", string(entity)).Msg("audit log insert failed")
	}
}

func (l *Logger) Created(ctx context.Context, entity sqlc.AuditEntity, entityID, performedBy, restaurantID pgtype.UUID, after any) {
	l.log(ctx, entity, entityID, performedBy, restaurantID, sqlc.AuditActionCREATE, nil, after)
}

func (l *Logger) Updated(ctx context.Context, entity sqlc.AuditEntity, entityID, performedBy, restaurantID pgtype.UUID, before, after any) {
	l.log(ctx, entity, entityID, performedBy, restaurantID, sqlc.AuditActionUPDATE, before, after)
}

func (l *Logger) Deleted(ctx context.Context, entity sqlc.AuditEntity, entityID, performedBy, restaurantID pgtype.UUID, before any) {
	l.log(ctx, entity, entityID, performedBy, restaurantID, sqlc.AuditActionDELETE, before, nil)
}

func marshalJSON(v any) []byte {
	if v == nil {
		return []byte("null")
	}
	b, err := json.Marshal(v)
	if err != nil {
		return []byte("null")
	}
	return b
}
